adapter/internal/sourcewatcher/sourcewatcher.go

```diff
@@ -0,0 +1,365 @@
+/*
+ *  Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package sourcewatcher
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/wso2/product-microgateway/adapter/config"
+	"github.com/wso2/product-microgateway/adapter/internal/api"
+	xds "github.com/wso2/product-microgateway/adapter/internal/discovery/xds"
+	"github.com/wso2/product-microgateway/adapter/internal/loggers"
+	"github.com/wso2/product-microgateway/adapter/internal/oasparser/model"
+	"github.com/wso2/product-microgateway/adapter/pkg/auth"
+	"github.com/wso2/product-microgateway/adapter/pkg/logging"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+)
+
+const (
+	apisArtifactDir string = "apis"
+	zipExt          string = ".zip"
+)
+
+var artifactsMap map[string]model.ProjectAPI
+
+// Start fetches the API artifacts at the startup and polls for changes from the remote repository
+func Start() {
+	conf, err := config.ReadConfigs()
+	if err != nil {
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Error reading configs : %s", err.Error()),
+			Severity: logging.BLOCKER,
+			ErrorCode: 2500,
+		})
+	}
+
+	retryInterval := conf.Adapter.SourceControl.RetryInterval
+
+	loggers.LoggerSourceWatcher.Info("Starting source watcher")
+	// Fetch the API artifacts at the startup
+	repository, err := fetchArtifacts()
+
+	// Retry fetching the API artifacts if the first attempt fails
+	for {
+		if err == nil {
+			break
+		} else {
+			time.Sleep(time.Duration(retryInterval) * time.Second)
+			loggers.LoggerSourceWatcher.Info("Retrying fetching artifacts from the remote repository")
+			repository, err = fetchArtifacts()
+		}
+	}
```

Comment:
Do we have a max retry count? As in what would happen the remote location never became alive?

Reply:
A max retry count wasn't specified. It keeps retrying until the repository is alive.

Reply:
Let's specify an max retry count too. This would save us from an infinite loop that could occur during an integration test failure too.

Reply:
Fixed with 92d923b